packages/server/modules/core/services/ratelimiter.ts

```diff
@@ -0,0 +1,344 @@
+import express from 'express'
+import Redis from 'ioredis'
+import {
+  getRedisUrl,
+  getIntFromEnv,
+  isTestEnv
+} from '@/modules/shared/helpers/envHelper'
+import {
+  BurstyRateLimiter,
+  RateLimiterAbstract,
+  RateLimiterMemory,
+  RateLimiterRedis,
+  RateLimiterRes
+} from 'rate-limiter-flexible'
+import { TIME } from '@speckle/shared'
+import { AuthContext } from '@/modules/shared/authz'
+import { BaseError } from '@/modules/shared/errors'
+import { getIpFromRequest } from '@/modules/shared/utils/ip'
+import { Options } from 'verror'
+
+export class RateLimitError extends BaseError {
+  static defaultMessage =
+    'You have sent too many requests. You are being rate limited. Please try again later.'
+  static code = 'RATE_LIMIT_ERROR'
+
+  rateLimitBreached: RateLimitBreached
+
+  constructor(
+    rateLimitBreached: RateLimitBreached,
+    message?: string | null | undefined,
+    options: Options | Error | undefined = undefined
+  ) {
+    super(message ?? RateLimitError.defaultMessage, options)
+    this.rateLimitBreached = rateLimitBreached
+  }
+}
+
+// typescript definitions
+
+type BurstyRateLimiterOptions = {
+  regularOptions: RateLimits
+  burstOptions: RateLimits
+}
+
+export type RateLimits = {
+  limitCount: number
+  duration: number
+}
+
+export enum RateLimitAction {
+  ALL_REQUESTS = 'ALL_REQUESTS',
+  USER_CREATE = 'USER_CREATE',
+  STREAM_CREATE = 'STREAM_CREATE',
+  COMMIT_CREATE = 'COMMIT_CREATE',
+  'POST /api/getobjects/:streamId' = 'POST /api/getobjects/:streamId',
+  'POST /api/diff/:streamId' = 'POST /api/diff/:streamId',
+  'POST /objects/:streamId' = 'POST /objects/:streamId',
+  'GET /objects/:streamId/:objectId' = 'GET /objects/:streamId/:objectId',
+  'GET /objects/:streamId/:objectId/single' = 'GET /objects/:streamId/:objectId/single',
+  'POST /graphql' = 'POST /graphql'
+}
+
+type RateLimiterOptions = {
+  [key in RateLimitAction]: BurstyRateLimiterOptions
+}
+
+interface RateLimitContext extends AuthContext {
+  ip: string
+}
+
+interface RequestWithContext extends express.Request {
+  context: RateLimitContext
```

Comment:
Preferably we should use type augmentations (see `type-augmentations` folder) to augment the global express.Request type so that we can get the context object anywhere.

We can mark the property as optional (`?`), but then we'll always have to check if its not undefined, so I think we should skip that, even if it makes writing code before the ctx object is initialized (which is a very small part of the project) a bit less type-safe